File: tests/refs/reverse_linked_list.cpp
// Reference C++ driver for tests/cases/reverse_linked_list.json.
// Treats the list as an array (per fixture schema) and reverses it.
#include <iostream>
#include <sstream>
#include <string>
#include <vector>
#include <algorithm>

struct Node { long long val; Node* next; };
std::vector<long long> reverseList(const std::vector<long long>& in) {
    Node* head = nullptr;
    for (auto it = in.rbegin(); it != in.rend(); ++it) head = new Node{*it, head};
    Node* prev = nullptr; Node* cur = head;
    while (cur) { Node* nx = cur->next; cur->next = prev; prev = cur; cur = nx; }
    std::vector<long long> out;
    while (prev) { out.push_back(prev->val); Node* nx = prev->next; delete prev; prev = nx; }
    return out;
}

int main() {
    std::string line, name; std::vector<long long> arr, expected;
    bool hi = false, he = false; int phase = 0, p = 0, f = 0;
    auto run = [&]() {
        if (name.empty() || !hi || !he) return;
        auto got = reverseList(arr);
        if (got == expected) { std::cout << "PASS reverse_linked_list :: " << name << "\n"; ++p; }
        else { std::cout << "FAIL reverse_linked_list :: " << name << "\n"; ++f; }
        name.clear(); arr.clear(); expected.clear(); hi = he = false; phase = 0;
    };
    while (std::getline(std::cin, line)) {
        if (line.empty()) { run(); continue; }
        std::istringstream iss(line); std::string tag; iss >> tag;
        if (tag == "CASE") { iss >> name; phase = 0; }
        else if (tag == "ARR") {
            std::size_t n; iss >> n;
            std::vector<long long> tmp(n);
            for (std::size_t i = 0; i < n; ++i) iss >> tmp[i];
            if (phase == 0) { arr = std::move(tmp); hi = true; phase = 1; }
            else { expected = std::move(tmp); he = true; }
        }
    }
    run();
    std::cout << "TOTAL: " << p << " passed, " << f << " failed\n";
    return f == 0 ? 0 : 1;
}

File: tests/refs/reverse_linked_list.rs
// Reference Rust driver for tests/cases/reverse_linked_list.json.
// Uses a safe, owned Box-based singly linked list. Reverses iteratively.
use std::io::{self, BufRead};

struct Node { val: i64, next: Option<Box<Node>> }

fn build(arr: &[i64]) -> Option<Box<Node>> {
    let mut head: Option<Box<Node>> = None;
    for &v in arr.iter().rev() { head = Some(Box::new(Node { val: v, next: head })); }
    head
}
fn to_vec(mut head: Option<Box<Node>>) -> Vec<i64> {
    let mut out = Vec::new();
    while let Some(n) = head { out.push(n.val); head = n.next; }
    out
}
fn reverse(mut head: Option<Box<Node>>) -> Option<Box<Node>> {
    let mut prev: Option<Box<Node>> = None;
    while let Some(mut node) = head {
        head = node.next.take();
        node.next = prev;
        prev = Some(node);
    }
    prev
}

fn main() {
    let stdin = io::stdin();
    let mut name = String::new();
    let mut arr: Vec<i64> = Vec::new(); let mut expected: Vec<i64> = Vec::new();
    let mut hi = false; let mut he = false; let mut phase = 0;
    let mut p = 0u32; let mut f = 0u32;
    let flush = |name: &mut String, arr: &mut Vec<i64>, expected: &mut Vec<i64>,
                 hi: &mut bool, he: &mut bool, phase: &mut i32, p: &mut u32, f: &mut u32| {
        if !name.is_empty() && *hi && *he {
            let got = to_vec(reverse(build(arr)));
            if got == *expected { println!("PASS reverse_linked_list :: {}", name); *p += 1; }
            else { println!("FAIL reverse_linked_list :: {}\n  expected: {:?}\n  got: {:?}", name, expected, got); *f += 1; }
        }
        name.clear(); arr.clear(); expected.clear(); *hi = false; *he = false; *phase = 0;
    };
    for line in stdin.lock().lines() {
        let line = line.unwrap_or_default();
        if line.is_empty() { flush(&mut name, &mut arr, &mut expected, &mut hi, &mut he, &mut phase, &mut p, &mut f); continue; }
        let mut parts = line.split_whitespace();
        let tag = parts.next().unwrap_or("");
        match tag {
            "CASE" => { name = parts.next().unwrap_or("").to_string(); phase = 0; }
            "ARR" => {
                let n: usize = parts.next().and_then(|s| s.parse().ok()).unwrap_or(0);
                let mut tmp: Vec<i64> = Vec::with_capacity(n);
                for _ in 0..n { if let Some(t) = parts.next() { if let Ok(v) = t.parse::<i64>() { tmp.push(v); } } }
                if phase == 0 { arr = tmp; hi = true; phase = 1; } else { expected = tmp; he = true; }
            }
            _ => {}
        }
    }
    flush(&mut name, &mut arr, &mut expected, &mut hi, &mut he, &mut phase, &mut p, &mut f);
    println!("TOTAL: {} passed, {} failed", p, f);
    std::process::exit(if f == 0 { 0 } else { 1 });
}
